hera_cal/lstbin_simple.py

```diff
@@ -0,0 +1,1740 @@
+        Shape ``(nbl, nlst_bins, nfreq, npol)``.
+    """
+    nlst_bins = len(data)
+    (_, nbl, nfreq, npol) = data[0].shape
+
+    for d, f, n in zip(data, flags, nsamples):
+        assert d.shape == f.shape == n.shape
+
+    # Do this just so that we can save memory if the call to this function already
+    # has allocated memory.
+
+    out_data = np.zeros((nbl, nlst_bins, nfreq, npol), dtype=complex)
+    out_flags = np.zeros(out_data.shape, dtype=bool)
+    out_std = np.ones(out_data.shape, dtype=complex)
+    out_nsamples = np.zeros(out_data.shape, dtype=float)
+
+    for lstbin, (d,n,f) in enumerate(zip(data, nsamples, flags)):
+        logger.info(f"Computing LST bin {lstbin+1} / {nlst_bins}")
+        
+        # TODO: check that this doesn't make yet another copy...
+        # This is just the data in this particular lst-bin.
+        if d.size:
+            (
+                out_data[:, lstbin], 
+                out_flags[:, lstbin], 
+                out_std[:, lstbin], 
+                out_nsamples[:, lstbin]
+            ) = lst_average(
+                d, n, f, mutable=mutable, 
+                sigma_clip_thresh=sigma_clip_thresh, 
+                sigma_clip_min_N=sigma_clip_min_N,
+                flag_below_min_N=flag_below_min_N,
+            )
+        else:
+            out_data[:, lstbin] = 1.0
+            out_flags[:, lstbin] = True
+            out_std[:, lstbin] = 1.0
+            out_nsamples[:, lstbin] = 0.0
+
+        
+    return out_data, out_flags, out_std, out_nsamples
+
+def _allocate_dnf(shape: tuple[int], d=0.0, f=0, n=0):
+    data = np.full(shape, d, dtype=complex)
+    flags = np.full(shape, f, dtype=bool)
+    nsamples = np.full(shape, n, dtype=float)
+    return data, flags, nsamples
+
+@profile
+def lst_average(
+    data: np.ndarray, nsamples: np.ndarray, flags: np.ndarray, 
+    flag_thresh: float = 0.7,
+    mutable: bool = False,
+    sigma_clip_thresh: float = 0.0,
+    sigma_clip_min_N: int=4,
+    flag_below_min_N: bool = False,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Compute statistics of a set of data over its zeroth axis.
+
+    The idea here is that the data's zeroth axis is "nights", and that each night is
+    at the same LST. However, this function is agnostic to the meaning of the first
+    axis. It just computes the mean, std, and nsamples over the first axis.
+
+    This function is meant to be used on a single element of a list returned by
+    :func:`simple_lst_bin`.
+
+    Parameters
+    ----------
+    data
+        The data to compute the statistics over. Shape ``(nnights, nbl, nfreq, npol)``.
+    nsamples
+        The number of samples for each measurement. Same shape as ``data``.
+    flags
+        The flags for each measurement. Same shape as ``data``.
+    flag_thresh
+        The fraction of times a baseline/frequency/pol must be flagged in order to
+        entirely flag the baseline/frequency/pol.
```

Comment:
I also hesitated over this. One thing I haven't quite landed on yet that I don't quite agree with you on is using "nights" here, because it's possible that later we use two time-integrations per night in each LST bin, then this doesn't correspond to nights, but maybe "integrations"?

So how about "The fraction of integrations within an LST bin for a given baseline (including pol) and channel that must be flagged in order for this function to flag that baseline-channel completely"?

Reply:
That works for me.